File: java/src/test/java/com/notifyservice/SendPipelineE2ETest.java
package com.notifyservice;

import static org.junit.jupiter.api.Assertions.assertEquals;

import com.notifyservice.api.model.NotificationChannel;
import com.notifyservice.api.model.Recipient;
import com.notifyservice.api.model.SendNotificationRequest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable;

@EnabledIfEnvironmentVariable(named = "NOTIFY_BASE_URL", matches = ".+")
@EnabledIfEnvironmentVariable(named = "NOTIFY_API_KEY", matches = ".+")
class SendPipelineE2ETest {

    @Test
    void sendReachesSentStatus() throws Exception {
        NotifyClient notify = new NotifyClient(System.getenv("NOTIFY_BASE_URL"), System.getenv("NOTIFY_API_KEY"));

        SendNotificationRequest req = new SendNotificationRequest()
                .channel(NotificationChannel.EMAIL)
                .recipient(new Recipient().address("alice@example.com").name("Alice"))
                .subject("Java SDK E2E")
                .htmlBody("<p>via java SDK over real HTTP</p>");

        var created = notify.notifications.sendNotification(req, null);
        String status = String.valueOf(created.getStatus());
        long deadline = System.currentTimeMillis() + 30_000;
        while (System.currentTimeMillis() < deadline
                && (status.equalsIgnoreCase("queued") || status.equalsIgnoreCase("sending"))) {
            Thread.sleep(300);
            status = String.valueOf(notify.notifications.getNotification(created.getId()).getStatus());
        }

        assertEquals("sent", status.toLowerCase());
    }
}

File: php/tests/SendPipelineE2ETest.php
<?php

declare(strict_types=1);

namespace NotifyService\Tests;

use NotifyService\Api\Model\NotificationChannel;
use NotifyService\Api\Model\Recipient;
use NotifyService\Api\Model\SendNotificationRequest;
use NotifyService\NotifyClient;
use PHPUnit\Framework\TestCase;

final class SendPipelineE2ETest extends TestCase
{
    public function testSendReachesSentStatus(): void
    {
        $baseUrl = getenv('NOTIFY_BASE_URL');
        $apiKey = getenv('NOTIFY_API_KEY');
        if ($baseUrl === false || $apiKey === false || $baseUrl === '' || $apiKey === '') {
            self::markTestSkipped('NOTIFY_BASE_URL/NOTIFY_API_KEY not set.');
        }

        $notify = new NotifyClient($baseUrl, $apiKey);
        $request = (new SendNotificationRequest())
            ->setChannel(NotificationChannel::EMAIL)
            ->setRecipient((new Recipient())->setAddress('alice@example.com')->setName('Alice'))
            ->setSubject('PHP SDK E2E')
            ->setHtmlBody('<p>via php SDK over real HTTP</p>');
        $created = $notify->notifications->sendNotification($request);

        $status = $created->getStatus()->value;
        $deadline = time() + 30;
        while (time() < $deadline && in_array(strtolower($status), ['queued', 'sending'], true)) {
            usleep(300000);
            $status = $notify->notifications->getNotification($created->getId())->getStatus()->value;
        }

        self::assertSame('sent', strtolower($status), "final status was {$status}");
    }
}
